File: backend/app/models/grievance_model.py
import uuid
from datetime import datetime, timezone
from typing import List, Optional
from sqlalchemy import Boolean, Float, ForeignKey, Integer, JSON, String, Text, DateTime
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database.session import Base


class ExtractionStatus:
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    NEEDS_VERIFICATION = "needs_verification"
    FAILED = "failed"


class GrievanceStatus:
    DRAFT = "draft"
    INTAKE_RECEIVED = "intake_received"
    CITIZEN_VERIFIED = "citizen_verified"
    GROUPING_CONFIRMED = "grouping_confirmed"
    AI_ANALYSIS_INTERVIEW = "ai_analysis_interview"
    UNDER_ANALYSIS = "under_analysis"
    FINALIZED = "finalized"
    AUTHORITY_RECOMMENDED = "authority_recommended"
    ADMINISTRATIVE_REVIEW = "administrative_review"
    FORWARDED = "forwarded"
    UNDER_PROCESSING = "under_processing"
    CLARIFICATION_REQUIRED = "clarification_required"
    RESOLVED = "resolved"
    CLOSED = "closed"


class IntakeMode:
    DIRECT_TEXT = "direct_text"
    OCR_HANDWRITTEN = "ocr_handwritten"
    VOICE_STT = "voice_stt"


class AttachmentType:
    HANDWRITTEN_PETITION = "handwritten_petition"
    SCANNED_DOCUMENT = "scanned_document"
    VOICE_RECORDING = "voice_recording"
    SUPPORTING_EVIDENCE = "supporting_evidence"


class InterviewStatus:
    NOT_REQUIRED = "not_required"
    NEEDS_INTERVIEW = "needs_interview"
    IN_PROGRESS = "in_progress"
    WAITING_FOR_CITIZEN = "waiting_for_citizen"
    COMPLETED = "completed"


class Grievance(Base):
    __tablename__ = "grievances"

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
        index=True,
    )
    grievance_number: Mapped[str] = mapped_column(
        String(50),
        unique=True,
        index=True,
        nullable=False,
    )
    citizen_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    title: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    intake_mode: Mapped[str] = mapped_column(
        String(50),
        default=IntakeMode.DIRECT_TEXT,
        nullable=False,
    )
    original_language: Mapped[str] = mapped_column(
        String(20),
        default="ml",
        nullable=False,
    )
    original_text: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    translated_text: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(
        String(50),
        default=GrievanceStatus.DRAFT,
        nullable=False,
    )
    priority: Mapped[str] = mapped_column(
        String(20),
        default="medium",
        nullable=False,
    )
    severity: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)
    summary: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    # Sprint 11 AI Processing fields
    ai_processing_status: Mapped[Optional[str]] = mapped_column(
        String(50), default="pending", nullable=True
    )
    ai_processed_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    ai_model: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    ai_error_message: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    confirmed_location: Mapped[Optional[dict]] = mapped_column(JSON, nullable=True)
    location_sources: Mapped[Optional[dict]] = mapped_column(JSON, nullable=True)
    category: Mapped[Optional[str]] = mapped_column(String(150), nullable=True)
    department_id: Mapped[Optional[str]] = mapped_column(String(150), nullable=True)
    assigned_official_id: Mapped[Optional[str]] = mapped_column(
        String(36),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    citizen = relationship("User", foreign_keys=[citizen_id], back_populates="grievances")
    assigned_official = relationship("User", foreign_keys=[assigned_official_id])
    attachments: Mapped[List["GrievanceAttachment"]] = relationship(
        "GrievanceAttachment",
        back_populates="grievance",
        cascade="all, delete-orphan",
        lazy="selectin",
    )
    audit_logs: Mapped[List["GrievanceAuditLog"]] = relationship(
        "GrievanceAuditLog",
        back_populates="grievance",
        cascade="all, delete-orphan",
        lazy="selectin",
    )
    analysis: Mapped[Optional["GrievanceAnalysis"]] = relationship(
        "GrievanceAnalysis",
        uselist=False,
        back_populates="grievance",
        cascade="all, delete-orphan",
        lazy="selectin",
    )
    issues: Mapped[List["GrievanceIssue"]] = relationship(
        "GrievanceIssue",
        back_populates="grievance",
        cascade="all, delete-orphan",
        lazy="selectin",
    )
    interview_questions: Mapped[List["GrievanceInterviewQuestion"]] = relationship(
        "GrievanceInterviewQuestion",
        back_populates="grievance",
        cascade="all, delete-orphan",
        lazy="selectin",
    )
    interview_responses: Mapped[List["GrievanceInterviewResponse"]] = relationship(
        "GrievanceInterviewResponse",
        back_populates="grievance",
        cascade="all, delete-orphan",
        lazy="selectin",
    )
    ai_runs: Mapped[List["GrievanceAIRun"]] = relationship(
        "GrievanceAIRun",
        back_populates="grievance",
        cascade="all, delete-orphan",
        lazy="selectin",
    )


class GrievanceIssue(Base):
    __tablename__ = "grievance_issues"

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
        index=True,
    )
    grievance_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("grievances.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    issue_number: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    category: Mapped[str] = mapped_column(String(150), nullable=False)
    subcategory: Mapped[Optional[str]] = mapped_column(String(150), nullable=True)
    severity: Mapped[str] = mapped_column(String(50), default="MEDIUM", nullable=False)
    priority: Mapped[str] = mapped_column(String(50), default="MEDIUM", nullable=False)
    status: Mapped[str] = mapped_column(String(50), default="OPEN", nullable=False)
    extracted_facts: Mapped[Optional[dict]] = mapped_column(JSON, nullable=True)
    interview_status: Mapped[str] = mapped_column(
        String(50), default=InterviewStatus.NOT_REQUIRED, nullable=False
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    grievance = relationship("Grievance", back_populates="issues")
    questions = relationship("GrievanceInterviewQuestion", back_populates="issue", cascade="all, delete-orphan")


class GrievanceInterviewQuestion(Base):
    __tablename__ = "grievance_interview_questions"

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
        index=True,
    )
    grievance_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("grievances.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    issue_id: Mapped[Optional[str]] = mapped_column(
        String(36),
        ForeignKey("grievance_issues.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )
    question: Mapped[str] = mapped_column(Text, nullable=False)
    question_type: Mapped[str] = mapped_column(String(50), default="TEXT", nullable=False)
    required: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    order_index: Mapped[int] = mapped_column(Integer, default=1, nullable=False)
    status: Mapped[str] = mapped_column(String(50), default="PENDING", nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    grievance = relationship("Grievance", back_populates="interview_questions")
    issue = relationship("GrievanceIssue", back_populates="questions")
    responses = relationship("GrievanceInterviewResponse", back_populates="question", cascade="all, delete-orphan")


class GrievanceInterviewResponse(Base):
    __tablename__ = "grievance_interview_responses"

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
        index=True,
    )
    question_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("grievance_interview_questions.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    issue_id: Mapped[Optional[str]] = mapped_column(
        String(36),
        ForeignKey("grievance_issues.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )
    grievance_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("grievances.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    response_text: Mapped[str] = mapped_column(Text, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    grievance = relationship("Grievance", back_populates="interview_responses")
    question = relationship("GrievanceInterviewQuestion", back_populates="responses")


class GrievanceAIRun(Base):
    __tablename__ = "grievance_ai_runs"

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
        index=True,
    )
    grievance_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("grievances.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    operation: Mapped[str] = mapped_column(String(100), nullable=False)
    model: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    status: Mapped[str] = mapped_column(String(50), default="PENDING", nullable=False)
    input_summary: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    output_data: Mapped[Optional[dict]] = mapped_column(JSON, nullable=True)
    error_message: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    completed_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    grievance = relationship("Grievance", back_populates="ai_runs")


class GrievanceAttachment(Base):
    __tablename__ = "grievance_attachments"

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
        index=True,
    )
    grievance_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("grievances.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    attachment_type: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
    )
    original_filename: Mapped[str] = mapped_column(String(255), nullable=False)
    mime_type: Mapped[str] = mapped_column(String(100), nullable=False)
    storage_path: Mapped[str] = mapped_column(String(500), nullable=False)
    file_size_bytes: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    raw_extracted_text: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    extraction_status: Mapped[str] = mapped_column(
        String(50),
        default=ExtractionStatus.PENDING,
        server_default=ExtractionStatus.PENDING,
        nullable=False,
    )
    extraction_confidence: Mapped[Optional[float]] = mapped_column(
        Float, nullable=True
    )
    extraction_engine: Mapped[Optional[str]] = mapped_column(
        String(100), nullable=True
    )
    extraction_error: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    extracted_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    grievance = relationship("Grievance", back_populates="attachments")


class GrievanceAuditLog(Base):
    __tablename__ = "grievance_audit_logs"

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
        index=True,
    )
    grievance_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("grievances.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    actor_id: Mapped[Optional[str]] = mapped_column(
        String(36),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    actor_role: Mapped[str] = mapped_column(
        String(50),
        default="citizen",
        nullable=False,
    )
    action_type: Mapped[str] = mapped_column(String(50), nullable=False)
    previous_state: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    new_state: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    remarks: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    grievance = relationship("Grievance", back_populates="audit_logs")
    actor = relationship("User")


class GrievanceAnalysis(Base):
    __tablename__ = "grievance_analysis"

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
        index=True,
    )
    grievance_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("grievances.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    extracted_entities: Mapped[Optional[dict]] = mapped_column(JSON, nullable=True)
    predicted_category: Mapped[Optional[str]] = mapped_column(String(150), nullable=True)
    legal_grounding_references: Mapped[Optional[dict]] = mapped_column(JSON, nullable=True)
    ai_explanation: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    grievance = relationship("Grievance", back_populates="analysis")

File: lib/models/grievance_model.dart
class AIDecisionSupportModel {
  final String suggestedDepartment;
  final String priority;
  final String reasoning;
  final List<String> keyFacts;
  final String statutoryRelevance;
  final String missingInformation;
  final String suggestedNextStep;
  final double confidenceScore;
  final String disclaimer;

  const AIDecisionSupportModel({
    required this.suggestedDepartment,
    required this.priority,
    required this.reasoning,
    required this.keyFacts,
    required this.statutoryRelevance,
    required this.missingInformation,
    required this.suggestedNextStep,
    required this.confidenceScore,
    required this.disclaimer,
  });

  factory AIDecisionSupportModel.fromJson(Map<String, dynamic> json) {
    var rawFacts = json['key_facts'] as List<dynamic>?;
    List<String> facts = rawFacts != null ? rawFacts.map((e) => e.toString()).toList() : [];

    return AIDecisionSupportModel(
      suggestedDepartment: (json['suggested_department'] ?? '') as String,
      priority: (json['priority'] ?? 'medium') as String,
      reasoning: (json['reasoning'] ?? '') as String,
      keyFacts: facts,
      statutoryRelevance: (json['statutory_relevance'] ?? '') as String,
      missingInformation: (json['missing_information'] ?? '') as String,
      suggestedNextStep: (json['suggested_next_step'] ?? '') as String,
      confidenceScore: ((json['confidence_score'] ?? 0.85) as num).toDouble(),
      disclaimer: (json['disclaimer'] ?? 'AI-assisted recommendation.') as String,
    );
  }
}

class GrievanceIssueModel {
  final String id;
  final String grievanceId;
  final int issueNumber;
  final String title;
  final String? description;
  final String category;
  final String? subcategory;
  final String severity;
  final String priority;
  final String status;
  final Map<String, dynamic>? extractedFacts;
  final String interviewStatus;
  final DateTime createdAt;
  final DateTime updatedAt;

  const GrievanceIssueModel({
    required this.id,
    required this.grievanceId,
    required this.issueNumber,
    required this.title,
    this.description,
    required this.category,
    this.subcategory,
    required this.severity,
    required this.priority,
    required this.status,
    this.extractedFacts,
    required this.interviewStatus,
    required this.createdAt,
    required this.updatedAt,
  });

  factory GrievanceIssueModel.fromJson(Map<String, dynamic> json) {
    return GrievanceIssueModel(
      id: (json['id'] ?? '').toString(),
      grievanceId: (json['grievance_id'] ?? '').toString(),
      issueNumber: (json['issue_number'] as num?)?.toInt() ?? 1,
      title: (json['title'] ?? 'Reported Issue').toString(),
      description: json['description']?.toString(),
      category: (json['category'] ?? 'General').toString(),
      subcategory: json['subcategory']?.toString(),
      severity: (json['severity'] ?? 'MEDIUM').toString(),
      priority: (json['priority'] ?? 'MEDIUM').toString(),
      status: (json['status'] ?? 'OPEN').toString(),
      extractedFacts: json['extracted_facts'] is Map<String, dynamic>
          ? json['extracted_facts'] as Map<String, dynamic>
          : null,
      interviewStatus: (json['interview_status'] ?? 'not_required').toString(),
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'].toString()) ?? DateTime.now()
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.tryParse(json['updated_at'].toString()) ?? DateTime.now()
          : DateTime.now(),
    );
  }
}

class GrievanceInterviewQuestionModel {
  final String id;
  final String grievanceId;
  final String? issueId;
  final String question;
  final String questionType;
  final bool required;
  final int orderIndex;
  final String status;
  final DateTime createdAt;

  const GrievanceInterviewQuestionModel({
    required this.id,
    required this.grievanceId,
    this.issueId,
    required this.question,
    required this.questionType,
    required this.required,
    required this.orderIndex,
    required this.status,
    required this.createdAt,
  });

  factory GrievanceInterviewQuestionModel.fromJson(Map<String, dynamic> json) {
    return GrievanceInterviewQuestionModel(
      id: (json['id'] ?? '').toString(),
      grievanceId: (json['grievance_id'] ?? '').toString(),
      issueId: json['issue_id']?.toString(),
      question: (json['question'] ?? '').toString(),
      questionType: (json['question_type'] ?? 'TEXT').toString(),
      required: json['required'] as bool? ?? true,
      orderIndex: (json['order_index'] as num?)?.toInt() ?? 1,
      status: (json['status'] ?? 'PENDING').toString(),
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'].toString()) ?? DateTime.now()
          : DateTime.now(),
    );
  }
}

class GrievanceModel {
  final String id;
  final String grievanceNumber;
  final String citizenId;
  final String? title;
  final String? description;
  final String intakeMode;
  final String originalLanguage;
  final String? originalText;
  final String? translatedText;
  final String status;
  final String priority;
  final String? severity;
  final String? summary;
  final String? aiProcessingStatus;
  final DateTime? aiProcessedAt;
  final String? aiModel;
  final String? aiErrorMessage;
  final String? category;
  final String? departmentId;
  final String? citizenName;
  final String? citizenPhone;
  final String? rawOcrText;
  final String? predictedDepartment;
  final String? assignedDepartment;
  final String? aiExplanation;
  final AIDecisionSupportModel? decisionSupport;
  final List<GrievanceAttachmentModel> attachments;
  final List<GrievanceAuditLogModel> auditLogs;
  final List<GrievanceIssueModel> issues;
  final List<GrievanceInterviewQuestionModel> interviewQuestions;
  final DateTime createdAt;
  final DateTime updatedAt;

  const GrievanceModel({
    required this.id,
    required this.grievanceNumber,
    required this.citizenId,
    this.citizenName,
    this.citizenPhone,
    this.title,
    this.description,
    required this.intakeMode,
    required this.originalLanguage,
    this.originalText,
    this.translatedText,
    required this.status,
    required this.priority,
    this.severity,
    this.summary,
    this.aiProcessingStatus,
    this.aiProcessedAt,
    this.aiModel,
    this.aiErrorMessage,
    this.category,
    this.departmentId,
    this.rawOcrText,
    this.predictedDepartment,
    this.assignedDepartment,
    this.aiExplanation,
    this.decisionSupport,
    this.attachments = const [],
    this.auditLogs = const [],
    this.issues = const [],
    this.interviewQuestions = const [],
    required this.createdAt,
    required this.updatedAt,
  });

  factory GrievanceModel.fromJson(Map<String, dynamic> json) {
    var rawLogs = json['audit_logs'] as List<dynamic>?;
    List<GrievanceAuditLogModel> parsedLogs = rawLogs != null
        ? rawLogs
            .map((item) =>
                GrievanceAuditLogModel.fromJson(item as Map<String, dynamic>))
            .toList()
        : [];

    var rawAtts = json['attachments'] as List<dynamic>?;
    List<GrievanceAttachmentModel> parsedAtts = rawAtts != null
        ? rawAtts
            .map((item) =>
                GrievanceAttachmentModel.fromJson(item as Map<String, dynamic>))
            .toList()
        : [];

    var rawIssues = json['issues'] as List<dynamic>?;
    List<GrievanceIssueModel> parsedIssues = rawIssues != null
        ? rawIssues
            .map((item) =>
                GrievanceIssueModel.fromJson(item as Map<String, dynamic>))
            .toList()
        : [];

    var rawQuestions = json['interview_questions'] as List<dynamic>?;
    List<GrievanceInterviewQuestionModel> parsedQuestions = rawQuestions != null
        ? rawQuestions
            .map((item) =>
                GrievanceInterviewQuestionModel.fromJson(item as Map<String, dynamic>))
            .toList()
        : [];

    AIDecisionSupportModel? ds;
    if (json['decision_support'] != null && json['decision_support'] is Map<String, dynamic>) {
      ds = AIDecisionSupportModel.fromJson(json['decision_support'] as Map<String, dynamic>);
    }

    return GrievanceModel(
      id: (json['id'] ?? '').toString(),
      grievanceNumber: (json['grievance_number'] ?? json['id'] ?? '').toString(),
      citizenId: (json['citizen_id'] ?? '').toString(),
      citizenName: json['citizen_name']?.toString(),
      citizenPhone: json['citizen_phone']?.toString(),
      title: json['title']?.toString(),
      description: json['description']?.toString(),
      intakeMode: (json['intake_mode'] ?? 'direct_text').toString(),
      originalLanguage: (json['original_language'] ?? 'ml').toString(),
      originalText: json['original_text']?.toString(),
      translatedText: json['translated_text']?.toString(),
      status: (json['status'] ?? 'draft').toString(),
      priority: (json['priority'] ?? 'medium').toString(),
      severity: json['severity']?.toString(),
      summary: json['summary']?.toString(),
      aiProcessingStatus: json['ai_processing_status']?.toString(),
      aiProcessedAt: json['ai_processed_at'] != null ? DateTime.tryParse(json['ai_processed_at'].toString()) : null,
      aiModel: json['ai_model']?.toString(),
      aiErrorMessage: json['ai_error_message']?.toString(),
      category: json['category']?.toString(),
      departmentId: (json['department_id'] ?? json['assigned_department'] ?? json['predicted_department'])?.toString(),
      rawOcrText: json['raw_ocr_text']?.toString(),
      predictedDepartment: json['predicted_department']?.toString(),
      assignedDepartment: json['assigned_department']?.toString(),
      aiExplanation: json['ai_explanation']?.toString(),
      decisionSupport: ds,
      attachments: parsedAtts,
      auditLogs: parsedLogs,
      issues: parsedIssues,
      interviewQuestions: parsedQuestions,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'].toString()) ?? DateTime.now()
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.tryParse(json['updated_at'].toString()) ?? DateTime.now()
          : DateTime.now(),
    );
  }

  /// Returns a clean, human-readable title for the grievance.
  String getDisplayTitle() {
    if (title != null && title!.isNotEmpty && !title!.startsWith('Handwritten Petition Intake:')) {
      return title!;
    }

    final sourceText = description ?? originalText ?? rawOcrText;
    if (sourceText != null && sourceText.trim().isNotEmpty) {
      final clean = sourceText.trim();
      final lines = clean.split('\n');
      for (final line in lines) {
        final trimmed = line.trim();
        if (trimmed.toLowerCase().startsWith('വിഷയം:') || trimmed.toLowerCase().startsWith('subject:')) {
          final subjectPart = trimmed.substring(trimmed.indexOf(':') + 1).trim();
          if (subjectPart.isNotEmpty) {
            return subjectPart.length > 55 ? '${subjectPart.substring(0, 55)}...' : subjectPart;
          }
        }
      }
      for (final line in lines) {
        final trimmed = line.trim();
        if (trimmed.isNotEmpty && !trimmed.toLowerCase().startsWith('പരാതി തീയതി')) {
          return trimmed.length > 55 ? '${trimmed.substring(0, 55)}...' : trimmed;
        }
      }
      return clean.length > 55 ? '${clean.substring(0, 55)}...' : clean;
    }

    if (title != null && title!.isNotEmpty) {
      return title!;
    }

    return 'Public Grievance Petition #${grievanceNumber.substring(0, grievanceNumber.length > 8 ? 8 : grievanceNumber.length)}';
  }

  /// Returns a short briefing / summary snippet for UI list cards.
  String? getShortBriefing() {
    if (summary != null && summary!.isNotEmpty) {
      return summary;
    }
    final text = description ?? originalText ?? rawOcrText;
    if (text != null && text.trim().isNotEmpty) {
      final clean = text.trim();
      return clean.length > 120 ? '${clean.substring(0, 120)}...' : clean;
    }
    return null;
  }
}

class GrievanceAuditLogModel {
  final String id;
  final String grievanceId;
  final String? actorId;
  final String actorRole;
  final String actionType;
  final String? previousState;
  final String? newState;
  final String? remarks;
  final DateTime createdAt;

  const GrievanceAuditLogModel({
    required this.id,
    required this.grievanceId,
    this.actorId,
    required this.actorRole,
    required this.actionType,
    this.previousState,
    this.newState,
    this.remarks,
    required this.createdAt,
  });

  factory GrievanceAuditLogModel.fromJson(Map<String, dynamic> json) {
    return GrievanceAuditLogModel(
      id: (json['id'] ?? '').toString(),
      grievanceId: (json['grievance_id'] ?? '').toString(),
      actorId: json['actor_id']?.toString(),
      actorRole: (json['actor_role'] ?? 'citizen').toString(),
      actionType: (json['action_type'] ?? '').toString(),
      previousState: json['previous_state']?.toString(),
      newState: json['new_state']?.toString(),
      remarks: json['remarks']?.toString(),
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'].toString()) ?? DateTime.now()
          : DateTime.now(),
    );
  }
}

class GrievanceAttachmentModel {
  final String id;
  final String grievanceId;
  final String attachmentType;
  final String originalFilename;
  final String mimeType;
  final String storagePath;
  final int? fileSizeBytes;
  final String? rawExtractedText;
  final String extractionStatus;
  final double? extractionConfidence;
  final String? extractionEngine;
  final String? extractionError;
  final DateTime? extractedAt;
  final DateTime createdAt;

  const GrievanceAttachmentModel({
    required this.id,
    required this.grievanceId,
    required this.attachmentType,
    required this.originalFilename,
    required this.mimeType,
    required this.storagePath,
    this.fileSizeBytes,
    this.rawExtractedText,
    required this.extractionStatus,
    this.extractionConfidence,
    this.extractionEngine,
    this.extractionError,
    this.extractedAt,
    required this.createdAt,
  });

  factory GrievanceAttachmentModel.fromJson(Map<String, dynamic> json) {
    return GrievanceAttachmentModel(
      id: (json['id'] ?? '').toString(),
      grievanceId: (json['grievance_id'] ?? '').toString(),
      attachmentType: (json['attachment_type'] ?? '').toString(),
      originalFilename: (json['original_filename'] ?? '').toString(),
      mimeType: (json['mime_type'] ?? '').toString(),
      storagePath: (json['storage_path'] ?? '').toString(),
      fileSizeBytes: (json['file_size_bytes'] as num?)?.toInt(),
      rawExtractedText: json['raw_extracted_text']?.toString(),
      extractionStatus: (json['extraction_status'] ?? 'pending').toString(),
      extractionConfidence: (json['extraction_confidence'] as num?)?.toDouble(),
      extractionEngine: json['extraction_engine']?.toString(),
      extractionError: json['extraction_error']?.toString(),
      extractedAt: json['extracted_at'] != null
          ? DateTime.tryParse(json['extracted_at'].toString())
          : null,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'].toString()) ?? DateTime.now()
          : DateTime.now(),
    );
  }
}

class NormalizedExtractionResultModel {
  final String sourceType;
  final String? sourceAttachmentId;
  final String originalLanguage;
  final String? extractedText;
  final String extractionStatus;
  final double? confidenceScore;
  final String engineName;
  final DateTime processedAt;
  final String? errorMessage;

  const NormalizedExtractionResultModel({
    required this.sourceType,
    this.sourceAttachmentId,
    required this.originalLanguage,
    this.extractedText,
    required this.extractionStatus,
    this.confidenceScore,
    required this.engineName,
    required this.processedAt,
    this.errorMessage,
  });

  factory NormalizedExtractionResultModel.fromJson(Map<String, dynamic> json) {
    return NormalizedExtractionResultModel(
      sourceType: (json['source_type'] ?? '').toString(),
      sourceAttachmentId: json['source_attachment_id']?.toString(),
      originalLanguage: (json['original_language'] ?? 'ml').toString(),
      extractedText: json['extracted_text']?.toString(),
      extractionStatus: (json['extraction_status'] ?? '').toString(),
      confidenceScore: (json['confidence_score'] as num?)?.toDouble(),
      engineName: (json['engine_name'] ?? '').toString(),
      processedAt: json['processed_at'] != null
          ? DateTime.tryParse(json['processed_at'].toString()) ?? DateTime.now()
          : DateTime.now(),
      errorMessage: json['error_message']?.toString(),
    );
  }
}
